File: day_4/puzzle_1.py
from pprint import pprint, pformat

puzzle_input = "254032-789860"
[range_start, range_end] = [int(x) for x in puzzle_input.split("-")]

# brute force method

valid_passwords = []

for n in range(range_start, range_end):
    # print(n)

    n_str = str(n)

    # test double digits
    has_double = False
    for i in range(1, len(n_str)):
        # print(i)
        if n_str[i] == n_str[i-1]:
            has_double = True

    if not has_double:
        continue

    # test increasing digits
    has_inc_digits = True
    for i in range(1, len(n_str)):
        if int(n_str[i]) < int(n_str[i-1]):
            has_inc_digits = False
            break
    
    if not has_inc_digits:
        continue

    valid_passwords.append(n)


print("valid passwords:", valid_passwords)
print("# valid passwords:", len(valid_passwords))

File: go/day_04/puzzle_1.go
package main

import (
	"fmt"
	"strconv"
)

// Utility Functions
////////////////////////////////////////////////////////////////////////////////

func check(err error, msg string) {
	if err != nil {
		panic(fmt.Sprintf("%s: %s", msg, err))
	}
}

// Main
////////////////////////////////////////////////////////////////////////////////

func main() {

	rangeStart, rangeEnd := 254032, 789860
	validCount := 0

	// iterate through range
	for i := rangeStart; i < rangeEnd; i++ {

		hasDoubleDigits := false
		hasIncreasingDigits := true
		prevDigit := 0

		// iterate over digits
		for idx, digitStr := range strconv.Itoa(i) {
			digit, _ := strconv.Atoi(string(digitStr))

			// skip first digit
			if idx == 0 {
				prevDigit = digit
				continue
			}

			// test double digits
			if digit == prevDigit {
				hasDoubleDigits = true
			}

			// test increasing digits
			if digit < prevDigit {
				hasIncreasingDigits = false
			}

			prevDigit = digit
		}

		// increment if valid password
		if hasDoubleDigits && hasIncreasingDigits {
			validCount++
		}

	}

	fmt.Printf("Found %v valid passwords.\n", validCount)

}
